netx/dialer/quicsaver_test.go

```diff
@@ -0,0 +1,207 @@
+	}
+	if sess != nil {
+		t.Fatal("expected nil sess here")
+	}
+	ev := saver.Read()
+	if len(ev) != 1 {
+		t.Fatal("expected a single event here")
+	}
+	if ev[0].Address != "www.google.com:443" {
+		t.Fatal("unexpected Address", ev[0].Address)
+	}
+	if ev[0].Duration <= 0 {
+		t.Fatal("unexpected Duration")
+	}
+	if !errors.Is(ev[0].Err, expected) {
+		t.Fatal("unexpected Err")
+	}
+	if ev[0].Name != errorx.QUICHandshakeOperation {
+		t.Fatal("unexpected Name")
+	}
+	if ev[0].Proto != "udp" {
+		t.Fatal("unexpected Proto")
+	}
+	if !ev[0].Time.Before(time.Now()) {
+		t.Fatal("unexpected Time")
+	}
+}
+
+func TestQUICSaverConnDialSuccess(t *testing.T) {
+	tlsConf := &tls.Config{
+		NextProtos: []string{"h3-29"},
+	}
+	saver := &trace.Saver{}
+	systemdialer := dialer.QUICSystemDialer{Saver: saver}
+
+	sess, err := systemdialer.DialContext(context.Background(), "udp", "216.58.212.164:443", "www.google.com:443", tlsConf, &quic.Config{})
+	if err != nil {
+		t.Fatal("unexpected error", err)
+	}
+	if sess == nil {
+		t.Fatal("unexpected nil session")
+	}
+	ev := saver.Read()
+	if len(ev) < 4 {
+		// it's a bit tricky to be sure about the right number of
+		// events because network conditions may influence that
+		t.Fatal("unexpected number of events")
+	}
+
+	last := len(ev) - 1
+	for idx := 1; idx < last; idx++ {
+		if ev[idx].Data == nil {
+			t.Fatal("unexpected Data")
+		}
+		if ev[idx].Duration <= 0 {
+			t.Fatal("unexpected Duration")
+		}
+		if ev[idx].Err != nil {
+			t.Fatal("unexpected Err")
+		}
+		if ev[idx].NumBytes <= 0 {
+			t.Fatal("unexpected NumBytes")
+		}
+		switch ev[idx].Name {
+		case errorx.ReadOperation, errorx.WriteOperation:
+		default:
+			t.Fatal("unexpected Name")
+		}
+		if ev[idx].Time.Before(ev[idx-1].Time) {
+			t.Fatal("unexpected Time")
+		}
+	}
+}
+
+func TestQUICHandshakeSaverSuccess(t *testing.T) {
+	nextprotos := []string{"h3-29"}
+	servername := "www.google.com"
+	tlsConf := &tls.Config{
+		NextProtos: nextprotos,
+		ServerName: servername,
+	}
+	saver := &trace.Saver{}
+	dlr := dialer.QUICHandshakeSaver{
+		Dialer: dialer.QUICSystemDialer{},
+		Saver:  saver,
+	}
+
+	sess, err := dlr.DialContext(context.Background(), "udp", "216.58.212.164:443", "www.google.com:443", tlsConf, &quic.Config{})
+	if err != nil {
+		t.Fatal("unexpected error", err)
+	}
+	if sess == nil {
+		t.Fatal("unexpected nil sess")
+	}
+	ev := saver.Read()
+	if len(ev) != 2 {
+		t.Fatal("unexpected number of events")
+	}
+	if ev[0].Name != "quic_handshake_start" {
+		t.Fatal("unexpected Name")
+	}
+	if ev[0].TLSServerName != "www.google.com" {
+		t.Fatal("unexpected TLSServerName")
+	}
+	if !reflect.DeepEqual(ev[0].TLSNextProtos, nextprotos) {
+		t.Fatal("unexpected TLSNextProtos")
+	}
+	if ev[0].Time.After(time.Now()) {
+		t.Fatal("unexpected Time")
+	}
+	if ev[1].Duration <= 0 {
+		t.Fatal("unexpected Duration")
+	}
+	if ev[1].Err != nil {
+		t.Fatal("unexpected Err", ev[1].Err)
+	}
+	if ev[1].Name != "quic_handshake_done" {
```

Comment:
Yeah, these strings maybe should just be defined once in errorx

Reply:
I did this in correspondence to _saver_test.go_ , seeing that, for instance, "tls_handshake_done" is also not defined in errorx.

Reply:
Yes, I've just checked `saver.go`. It seems only the `tls_handshake_{start,done}` events do not have a corresponding name inside of `errorx.go`. I _think_ I understand what is going on. We have operations, whose name is defined in `errorx.go` and then we have events. Some events have the same name of operations, so we use the names in `errorx` while some others do not. This is the case of `"tls_handshake_done"`. It may be useful to have all these names in the same place to avoid typing errors, nonetheless. But maybe we can do this at a later time.

Reply:
yes, if so, it should be changed for TCP and QUIC both, I think

Reply:
I agree with you. This does not seem a blocker to me. Let's do nothing for now. The code is already consistent as is.